### PR TITLE
statement-distribution: support inactive local validator in grid

### LINKED ISSUE #1437: Test: new statement distribution functions when max_validators is set
https://github.com/paritytech/polkadot-sdk/pull/1436/commits/35be89f22da224f925d23757f372e07e2122fa64 in #1436 fixed this bug, where the grid topology would be incomplete if and only if `max_validators` was set (and the `validators` in the session-info had a smaller size than the `discovery_keys`).

This should be tested in the subsystem-wide tests, ensuring that the grid functions correctly even when both of the following are true:
  1. the local node has no key present in `validators` in the session info
  2. the `NewGossipTopology` message specifies an index of `validators.len() + 1`

### PR DESCRIPTION
**Pull request by slumber**
Fixes #1437

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3697704

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4194612

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4250800

**Comment by Sophia-Gold** [comment]
The implementation makes sense to me and the `inactive_local_participates_in_grid` test meets the criteria in #1437. 

The one thing I wonder is since I don't know how the bug this is meant to fix presents, should we try running this on Versi with `max_validators` set?

File: polkadot/node/network/statement-distribution/src/v2/tests/grid.rs [hunk 1]
```diff
@@ -1832,12 +1835,12 @@ fn advertisements_rejected_from_incorrect_peers() {
...
-		let v_a = target_group_validators[0];
-		let v_b = target_group_validators[1];
-		let v_c = other_group_validators[0];
-		let v_d = other_group_validators[1];
+		let other_group_validators = state.group_validators(local_group_index, true);
+		let target_group_validators = state.group_validators(other_group, true);
+		let v_a = other_group_validators[0];
+		let v_b = other_group_validators[1];
+		let v_c = target_group_validators[0];
+		let v_d = target_group_validators[1];
```
Comment (mrcnski) [inline polkadot/node/network/statement-distribution/src/v2/tests/grid.rs:1843]: Why was this changed? It doesn't match the comment below anymore.

**Review by mrcnski — APPROVED** [review]
Looks sensible and addresses the issue. Since I'm rusty with this code I would have liked to see more tests e.g. in `tests/cluster` and `tests/requests`. Also left some follow-ups.

One thing I noticed is that statement-distribution is highly branchy which is a bit exacerbated by this PR. This subsystem would benefit from code coverage statistics. Not that we should test every branch necessarily, but it would help us at least identify which branches are/aren't covered by tests already.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4328856

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4332655

**Review by eskimor — APPROVED** [review]
Looks ok as far as I can tell. Some logic is quite confusing, some cleanup would be beneficial at some point. (In part, I believe there are tickets for this already)
